src/polly/src/pollable.rs

```diff
@@ -13,6 +13,10 @@ pub struct Pollable {
 
 /// Wrapper for file descriptors.
 impl Pollable {
+    pub fn new(rawfd: RawFd) -> Pollable {
```

Comment:
I think we want to avoid adding this method because it's potentially dangerous to create a `Pollable` directly from a `RawFd` which is just an integer.  If we really need to do that, we can pass a borrowed `File` constructed via `File::from_raw_fd` (which is `unsafe`) to `Pollable::from`.

Reply:
Good catch on needing to mark this `unsafe` one way or another!

However, I don't see any point in doing the hacky
> pass a borrowed `File` constructed via `File::from_raw_fd` (which is `unsafe`) to `Pollable::from`

If the user wanted to pass some random integer instead of a valid `RawFd` they could still pass it to `File::from_raw_fd`.

I believe the idiomatic approach would be to implement the `unsafe std::os::unix::io::FromRawFd` trait for `Pollable`. What do you think?

Reply:
I agree with implementing `std::os::unix::io::FromRawFd`  for Pollable. 👍

Reply:
Implementing `FromRawFd` for `Pollable` is a bit tricky because the semantics of the `from_raw_fd` method involve taking exclusive ownership of the file descriptor and handling cleanup. I'm leaning towards keeping the current safe & minimal `Pollable` interface as is, and letting the user decide which safe wrapper that implements `AsRawFd` is appropriate when dealing with a actual raw fd (usually `File` I would assume).

Reply:
Hmm, you are right. But the problem still doesn't have a solution.

Using anything that involves the `FromRawFd trait` requires that the newly created object consumes ownership of the specified file descriptor, like you mentioned. This is incompatible with the `epoll` model where a FD is shared between the owner and epoll.

We could move the full EventManager interface to work with `EventFd`s so that we can clone them, but otherwise using `FromRawFd` won't work.

Reply:
Considering that we are relying on `epoll` to do the right thing and automatically unregister the rawfd if its owner closes it, I suggest we go back to my original proposal and simply mark that function `unsafe`.

Any other ideas are welcome.

Reply:
I’ll think some more on this, maybe we can avoid it altogether..

Reply:
I like keeping the interface as is (only accepting things that implement `AsRawFd`) because it remains simple and we encourage users not to use `RawFd`s without wrapping them in a safe abstraction first (which should also take care of the lifetime and cleanup aspects). If this is not something we can immediately agree on, I'm totally fine with proceeding in any other way for now and finishing the discussion at a later time (such as when upstreaming stuff to rust-vmm).

Reply:
I'm trying to find a way around this. The problem I'm facing is that `Serial` uses a `Box<dyn ReadableFd>` which doesn't implement `AsRawFd` and its content isn't sized and cannot be used with the current `Pollable::from` interface.

`ReadableFd` is `io::Read + AsRawFd`, but `Box<ReadableFd>` obviously doesn't implement the traits itself. Could I maybe implement `AsRawFd` for `Box<ReadableFd>`? I will give that a go and also think of other options.

Reply:
Welp, this looks difficult to handle with today's Rust. Implementing `AsRawFd`for `Box<something>` is prob not possible because of the orphan rule. I guess we can just move forward with the `unsafe new` alternative.

Reply:
Found a way around it, therefore I've removed the `unsafe Pollable::new` interface. Please take a look.